ml4gw/transforms/whitening.py

```diff
@@ -0,0 +1,253 @@
+            self.nfft = min(8 * (self.ntaps - 1), kernel_size)
+            # TODO: remove this error once the above
+            # implementation is optimized
+            raise NotImplementedError(
+                "An optimal torch implementation of whitening for short "
+                "filter padding is not complete. Use a larger value of pad."
+            )
+
+    def fit(
+        self,
+        kernel_length: float,
+        fftlength: Optional[float] = None,
+        highpass: Optional[float] = None,
+        sample_rate: Optional[float] = None,
+        **backgrounds: Background,
+    ) -> None:
+        """Compute a time domain filter from background
+
+        Args:
+            kernel_length:
+                The length in seconds of the timeseries data
+                on which this transform will be applied after
+                fitting.
+            fftlength:
+                If background data is passed as timeseries data,
+                the length of the FFT to use to compute the PSD
+                of the background data, in seconds. If left as `None`,
+                will default to `1 / kernel_length`.
+            highpass:
+                Cutoff high-pass frequency for the frequency response
+                of the fit time domain filter, in Hz. If left as `None`,
+                frequency response will be the inverse of the background
+                PSD across all bins.
+            sample_rate:
+                If background ata is passed as timeseries data,
+                the rate at which it is sampled. If left as `None`,
+                will default to the sample rate of the data on which
+                this transform is meant to be applied.
+            **chanels:
+                Background data to use to fit the whitening time
+                domain filter, whose frequency response in each
+                channel will be the inverse of the corresponding
+                background data's PSD. Should be passed in the
+                same order these channels are expected to fall
+                along the channel dimension of the data this
+                transform is called on. Can be passed either as
+                time domain or frequency domain data.
+        """
+
+        if len(backgrounds) != self.num_channels:
+            raise ValueError(
+                "Expected to fit whitening transform on {} background "
+                "timeseries, but was passed {}".format(
+                    self.num_channels, len(backgrounds)
+                )
+            )
+
+        self._check_kernel_length(kernel_length)
+        df = 1 / kernel_length
+        ncorner = int(highpass / df) if highpass else 0
+
+        tdfs = np.zeros((self.num_channels, 1, self.ntaps - 1))
+        for i, (channel, x) in enumerate(backgrounds.items()):
+            psd = normalize_psd(
+                x, df, self.sample_rate, sample_rate, fftlength
+            )
+            if (psd == 0).any():
```

Comment:
See above comment. Maybe in the future `normalize_psd` can handle the logic of rescaling data so that we don't run into this precision issue.